lib/IDE/Refactoring.cpp

```diff
@@ -3526,6 +3526,197 @@ performChange() {
+        : Printer(Printer) {}
+
+  bool walkToDeclPre(Decl *D) override {
+    if (auto valueD = dyn_cast<ValueDecl>(D)) {
+      if (!valueD->isSynthesized()) {
+          return true;
+      }
+      SmallString<32> Scratch;
+      auto name = valueD->getName().getString(Scratch);
+      // Print all synthesized enums,
+      // since Codable can synthesize multiple enums (for associated values).
+      auto shouldPrint = isa<EnumDecl>(valueD) ||
+        name == "init(from:)" ||
+        name == "encode(to:)";
+      if (!shouldPrint) {
+        // Some other synthesized decl that we don't want to print.
+        return false;
+      }
+    } else {
+        return false;
+    }
+    
+    Printer.printNewline();
+
+    if (auto enumDecl = dyn_cast<EnumDecl>(D)) {
+      // Manually print enum here, since we don't want to print synthesized functions.
+      Printer << "enum " << enumDecl->getNameStr();
+      PrintOptions Options;
+      Options.PrintSpaceBeforeInheritance = false;
+
+      enumDecl->printInherited(Printer, Options);
+      Printer << " {";
+      for (Decl *EC : enumDecl->getAllElements()) {
+        Printer.printNewline();
+        Printer << "  ";
+        EC->print(Printer, Options);
+      }
+      Printer.printNewline();
+      Printer << "}";
+      return false;
+    }
+    
+    SmallString<256> Buffer;
+    {
+      llvm::raw_svector_ostream OS(Buffer);
+      PrintOptions Options;
+      Options.SynthesizeSugarOnTypes = true;
+      Options.FunctionDefinitions = true;
+      Options.VarInitializers = true;
+      Options.PrintExprs = true;
+      Options.TypeDefinitions = true;
+      Options.ExcludeAttrList.push_back(DAK_HasInitialValue);
+
+      Printer.printNewline();
+      D->print(Printer, Options);
+    }
+    
+    Printer << Buffer.str();
+
+    return false;
+  }
+};
+
+std::string AddCodableContext::
+getInsertionText(const ResolvedCursorInfo &CursorInfo,
+                            SourceManager &SM) {
+  StringRef ExtraIndent;
+  StringRef CurrentIndent =
+    Lexer::getIndentationForLine(SM, getInsertStartLoc(), &ExtraIndent);
+  std::string Indent;
+  if (getInsertStartLoc() == StartLoc) {
+    Indent = (CurrentIndent + ExtraIndent).str();
+  } else {
+    Indent = CurrentIndent.str();
+  }
+
+  std::string Buffer;
+  llvm::raw_string_ostream OS(Buffer);
+  ExtraIndentStreamPrinter Printer(OS, Indent);
+  Printer.printNewline();
+  SynthesizedCodablePrinter Walker(Printer);
+  DC->getAsDecl()->walk(Walker);
+  return Buffer;
+}
+
+AddCodableContext AddCodableContext::
+getDeclarationContextFromInfo(ResolvedCursorInfo Info) {
+  if (Info.isInvalid()) {
+    return AddCodableContext();
+  }
+  if (!Info.IsRef) {
+    if (auto *NomDecl = dyn_cast<NominalTypeDecl>(Info.ValueD)) {
+      return AddCodableContext(NomDecl);
+    }
+  }
+  // FIXME: support extensions?
+  // (would need to get synthesized nodes from the main decl,
+  // and only if it's in the same file?)
```

Comment:
In general I think it would be nice to support adding the synthesized code to the extension, I imagine that it would be relatively common.

Reply:
I was hoping to get that support for free, but the synthesized nodes aren't in the extension. It might be trivial to get them from the main decl, but it's just enough trouble that I'd like to keep it as a follow up task.

Reply:
I assume they're on the main decl, which is trivial to get from the extension (but I haven't checked). Happy to leave as separate, though in that case I'd just leave a `TODO: Add extension support` or similar.

Reply:
Is the `FIXME` I added not enough? Or do you prefer if it really was a `TODO` instead?

Reply:
I mostly meant remove the `/**/` part. I tend to use "TODO" as "not doing this thing now" and "FIXME" as "fix before merging" but there's no convention there so use whatever you prefer.